### ISSUE TITLE
background color break after update

### ISSUE THREAD

**Issue by Saul-Mirone** [dracula/vim#65]
Hi, I'm using vim in iTerm on macOS. My iterm theme is dracula as well. It works well for years.
I update dracula today and my iterm color is so werid，and the background image disappear, what happend?

## BEFORE I UPDATE
![that](https://user-images.githubusercontent.com/10047788/38075312-8798d560-3364-11e8-9bba-abc9ec423a09.png)

## AFTER I UPDATE
![this](https://user-images.githubusercontent.com/10047788/38074787-9ec2615e-3362-11e8-94ef-fe4825584a5e.png)

**Comment by Saul-Mirone** [comment]
some hightlight break too, look at the arrows.
## BEFORE
![image](https://user-images.githubusercontent.com/10047788/38076599-ab0ab7c6-3368-11e8-854f-9c6adee485f3.png)
## AFTER
![image](https://user-images.githubusercontent.com/10047788/38076572-92698864-3368-11e8-9f3a-927aa772389f.png)

**Comment by Saul-Mirone** [comment]
visual mod breaks too

**Comment by Saul-Mirone** [comment]
JSX is the worest.....
![image](https://user-images.githubusercontent.com/10047788/38076827-65403846-3369-11e8-96b2-d9c071e49260.png)

**Comment by braddielman** [comment]
This is happening to me too, but with the terminal app. Background is dark grey instead of the default Dracula background.

**Comment by dsifford** [comment]
Fixes for @Saul-Mirone: Your terminal does not support italics. Or, if it does, your terminfo is not correct (which is why the italics are appearing inverted).

Try disabling the italics with `let g:dracula_italic = 0` in your vimrc

Fixes for @braddielman: Add `set termguicolors` to your vimrc.

**Comment by dsifford** [comment]
@Saul-Mirone  -- Re: Background image. To get that back you'll need to add a reset to the `Normal` highlight group.

so, for example: `highlight Normal cterm=NONE term=NONE` 

I'm not going to revert back to that because you and potentially a few others out there are in the minority when it comes to wanting a transparent background.

**Comment by dsifford** [comment]
@braddielman: Terminal App might not suffice for the fix that I suggested above as I'm not sure that that terminal supports truecolor.

If that's the case, you're seeing the gray because that's the closest 256 color fallback. See #60

**Comment by braddielman** [comment]
@dsifford - Yeah, using `set termguicolors` sets the background to bright yellow. The theme was working prior to the update, however. It's cool though since I mainly use MacVim.

**Comment by rivyx** [comment]
I'm on iTerm2 and my background color has also changed to match the airline gray. It was working until I did a vim-plug PlugUpdate a few minutes ago...

**Comment by rivyx** [comment]
`set termguicolors` seems to fix the issue outside of tmux, but when in tmux it looks like it completely breaks syntax highlighting

**Comment by dsifford** [comment]
@rivyx Haha, yeah I'm a tmux'er myself and there's a few things that you need to do to make it play nicely...

Can you confirm that you have the truecolor overrides in your tmux.conf _and_ the correct terminfo file is being read by tmux?

**Comment by rivyx** [comment]
the terminfo is the same for both my iTerm2 and my tmux (xterm-256color) and in an effort to keep my dotfiles portable I'd rather not implement truecolor overrides...just from a quick google it looks like a bit more overhead than I'm comfortable with

**Comment by dsifford** [comment]
Adding truecolor support with iterm should be no more than just adding a single line to your `tmux.conf`.

Adding a terminfo fix is only relevant if you want to make italics work correctly.

**Comment by dsifford** [comment]
For example, here's the line from my [tmux.conf](https://github.com/dsifford/.dotfiles/blob/master/tmux.conf#L11).

The relevant portion of that line with respect to truecolor is the `,*:Tc`.

**Comment by Saul-Mirone** [comment]
@dsifford Thanks for response. I tried disabling the italics, but the highlight and visual mod still not works correctly
![image](https://user-images.githubusercontent.com/10047788/38121232-9cbcb226-3400-11e8-87ea-c7c1a4ff420e.png)
(I select the "$ char '0'" using the visual mod but it seems still can't work, and the Integer highlight seems wrong)

**Comment by Saul-Mirone** [comment]
I forked and checkout to the commit [1f44625](https://github.com/dracula/vim/commit/1f44625290f0f1488f7056e166f69ee6f171d273)
Everything works fine now.
---
For Dracula, What important to me is that **it can be used together with it's iterm and zsh themes. So I can do every thing in terminal**. (I use solarized in gui vim). 
Thanks for the guys who contribute to this awesome theme, but these day's update seems break the most important advantage for some guys like me, so I have to use an old version managed by myself. 
I write this because I think it maybe useful for you guys to maintain this project.

**Comment by d3xru8** [comment]
I'm having @Saul-Mirone's issue too. Some keywords have cyan highlighting. And when I do a `:source %` from within vim, I get this output:
![screen shot 2018-03-30 at 12 35 00 pm](https://user-images.githubusercontent.com/9398261/38124776-01601a90-3417-11e8-88eb-c128773bf114.png)

**Comment by larshisken** [comment]
I'm using Vim8 in iTerm2 3.1.5 (also with dracula theme), using the following settings in this exact order fixed both issues for me:

```
let g:dracula_italic = 0
colorscheme dracula
highlight Normal ctermbg=None
```

**Comment by dsifford** [comment]
@larshisken Bingo!

Thanks for centralizing everything for people who are reading this thread.

Again, I'll likely add in a flag to disable the `ctermbg` in the `Normal` group when I get some free time.

@Saul-Mirone You're more than welcome to maintain a fork, but again the entirety of your issue boils down to having ctermbg set. So you're sort of creating more work for yourself.

If you can hold off for now, just add in the fix that @larshisken has above and I'll add in that flag soon.

**Comment by rivyx** [comment]
@dsifford So I got it working fine on my local instance of tmux inside of iTerm2 now that truecolors has been enabled and I've changed my terminfo to support italics and truecolors. However, I am now getting a seemingly related issue on Konsole (Kubuntu 16.04) inside of tmux (with terminfo and truecolors properly set).

![image](https://user-images.githubusercontent.com/16969699/38126236-3b0c8620-33bd-11e8-9fa7-cedecbefa1be.png)

**Also, I like using this theme on remote servers and I definitely don't always have the option of setting up truecolor overrides on there.**

Honestly, this update has really made working with this package a lot more difficult. I don't know what was wrong with the previous background color configuration, but until this issue is fixed/**truecolors are optional**, I am going to fork the project from a previous version...

**Comment by dsifford** [comment]
What's the unrelated issue you're getting?

Also, re: remote servers -- I adopted the same fallback mechanism for that that gruvbox uses: https://github.com/dracula/vim/blob/master/colors/dracula.vim#L28

If in the case that the remote server passes the fallback, setting the flag to disable ctermbg should make the theme work fine on remote servers.

---

Side note: I'm receiving a lot of flack here because of the update. I'll say this again: Nobody is stopping you from forking this and maintaining your own implementation that satisfies whatever look you're going for. The update was done for two reasons: 

1. The colors were out of specification.
2. The theme was difficult to maintain and difficult to extend with additional highlight groups.

I can't make everybody happy, so as a second best, I'm forced to make changes based on what a majority of users would agree upon. 

Thanks for understanding.

**Comment by rivyx** [comment]
Now getting the same issue inside of tmux on local Konsole (Kubuntu 16.04) and while inside of tmux on a remote server. This screenshot is from vim running in tmux on a remote server inside of iTerm2. It is the same problem from my previous comment (that screenshot was from my Kubuntu machine). It looks like the background color is being set correctly everywhere except the text characters (they have a gray highlight).

For Konsole, this problem persists even with truecolors set. For the remote server, I have ctermbg disabled and am getting the same problem.

![image](https://user-images.githubusercontent.com/16969699/38146210-e50c6c00-3401-11e8-9667-e16e745cd876.png)

**Comment by dsifford** [comment]
@rivyx The issue you're describing above is likely related to the ostensible fact that you have `termguicolors` indiscriminately enabled.

To get it to play well for all situations, you need to add a check in your vimrc that checks the environment you're in and then enables `termguicolors` based on that. If `termguicolors` is not enabled, then you should also (for now, until I get a chance to add the flag), add in a downgrade of the `Normal` highlight group so that `ctermbg=NONE`

Hope that makes sense.

**Comment by rivyx** [comment]
@dsifford Yes, I figured I realized just as you were commenting it seems. Also, sorry if I sound ungrateful, I do appreciate the work you've been putting in as well as all the help you've been providing for everyone. It was just very frustrating for a little while which I'm sure you get.

That being said, if it is possible to somehow tell dracula to set these settings automatically based on the current environment, that would be great! Thanks again...I luckily did not have to fork haha.

**Comment by Saul-Mirone** [comment]
@dsifford I tried @larshisken 's solution but it didn't work for me.

**Comment by dsifford** [comment]
Sorry for the delayed response.

@rivyx Re: "if it is possible to somehow tell dracula to set these settings automatically based on the current environment" --- I agree that would be ideal, but there's just too many variables in play to be able to do that effectively (terminal, os, kernel version, unpredictable terminfo across similar machines, etc). No way that I'm aware of to be able to reliably catch all of that.

Re: "Updating the instructions at https://draculatheme.com/vim/" -- @benknoble did a pretty amazing job with the help docs which are now available with `:h dracula` -- I think that might be something that we could potentially put in there. PRs accepted.

Glad to hear you were able to get everything sorted without having to fork!

@Saul-Mirone Refresh me again on what issues you have that are still unresolved? The same as your original comment or just a portion of that?

Can you link me to your vimrc? Also, can you provide some info related to your environment that you're on (terminal, os, vim/neovim/gvim, etc).. I understand you're using tmux. Have you tried terminal-override fixes?

**Comment by benknoble** [comment]
Hi all, sorry for leaping into the thread really late, just want to add a couple of things I saw.

1. I use Terminal.App on macos. It does *not*, I repeat, __not__, have truecolor support at this time
2. I have had *0* issues using the colorscheme since the update
 - I did set `g:dracula_italic=0` *before setting the colorscheme!*
 - I might turn off the background highlighting for Normal, since my term's background is setup to use Dracula colors (I maintain the Terminal.app theme) and the grey is kind of weird (but not impossible)
 - I guard against colorscheming twice in my vimrc, since that doesn't usually make sense and can ruin things like markology highlighting groups.
3. The entirety of my tmux/vim play-nice setup is using `screen-256color` in tmux and `xterm-256color` elsewhere; vim takes care of the rest automagically.
4. Truecolors has so far been optional ? That is, I do not (and should not) set `termguicolors`, and I experience no issues.

P.S. @dsifford , I haven't gone through the Terminal theme yet to see if it needs to be updated re: the spec.

When the new `ctermbg` flag gets added, let me know if you need help editing the doc file. It sounds like it could use an a explanation of what to with `termguicolors`, but I'm not in a position to author that sort of explanation.

**Comment by hacker1db** [comment]
I have same issue as above ^ @Saul-Mirone

**Comment by dsifford** [comment]
@Saul-Mirone @hacker1db 

Try setting `let g:dracula_colorterm = 0` to disable background fills.

**Comment by hacker1db** [comment]
@dsifford when I use tmux with that flag set it, the colorscheme loads as expected but when you use vim in just a standard terminal the theme does not load properly.

**Comment by benknoble** [comment]
@hacker1db How so? I have no bg issues in/outside of tmux now

**Comment by dsifford** [comment]
@hacker1db Yeah, you'll have to write in a conditional in your vimrc to apply that flag only why tmux is loaded.

so, basically...

```vim
if exists($TMUX)
  let g:dracula_colorterm = 0
endif
```

**Comment by dsifford** [comment]
@benknoble I think in @hacker1db's case, when tmux is not loaded his terminal background is black, where yours likely already has the dracula colors

**Comment by dsifford** [comment]
@hacker1db Can I assume you have the overrides set appropriately and you're calling `colorscheme dracula` somewhere in your init scripts after setting those overrides?

Do you have a link to an updated version of your startup scripts?

What terminal are you running?

**Comment by dsifford** [comment]
@hacker1db Looking at your init.vim

Are these lines correct?

https://github.com/hacker1db/Dotfiles/blob/dev_branch/config/nvim/init.vim#L39
https://github.com/hacker1db/Dotfiles/blob/dev_branch/config/nvim/init.vim#L42

**Comment by hacker1db** [comment]
@dsifford Sorry, can you please clarify your question are you thinking there not correct?

**Comment by dsifford** [comment]
Sure.. 

What I mentioned may or may not be correct based on your intentions, so I'm not sure. One thing I'm noticing is that you're explicitly forcing 256 color on all terminals and you are only setting "termguicolors" when not in tmux. Is that what you intended?

**Comment by hacker1db** [comment]
I guess, I was setting "termguicolors" to see if it would fix the Dracula theme when not in tmux.. as it works now when in tmux but when your not in tmux the color scheme does not apply but you manually apply it and all is well.

**Comment by dsifford** [comment]
Sorry for my slowness -- can you clarify again just once more what exactly is broken in non-tmux terminals?

Just want to make sure we're on the same page here.

**Comment by k-f-** [comment]
Updated the plug-in today and the background color changed to the issue folks are having here.
fixed with `let g:dracula_colorterm = 0` however the line number gutter no longer has a contrasting color.

**Comment by hacker1db** [comment]
@dsifford the colorscheme does not apply fully when not running tmux

**Comment by hacker1db** [comment]
Correct, I can't tell for sure if its affecting anything else but thats the most obvious issue. @dsifford

**Comment by dsifford** [comment]
Would you mind reporting back with the output of `:verbose hi NonText` in the affected terminal?

**Comment by dsifford** [comment]
Also, can you run `:set foldcolumn=3` and let me know if that is affected as well?

**Comment by hacker1db** [comment]
<img width="473" alt="screen shot 2018-04-06 at 2 56 20 pm" src="https://user-images.githubusercontent.com/6212228/38446047-e32a5e34-39aa-11e8-998d-4ca3ca220cfb.png">



<img width="288" alt="screen shot 2018-04-06 at 2 56 52 pm" src="https://user-images.githubusercontent.com/6212228/38446048-e34b141c-39aa-11e8-83b3-283b0ef67cd6.png">
@dsifford  here is the info you asked for let me know how else I can help is anyone else besides me having this issue? I wonder if its something I did in my config that is causing this issue...

**Comment by dsifford** [comment]
@hacker1db Thanks for that.

On my machine, the output is

```
:verbose hi NonText
NonText        xxx ctermfg=12 gui=bold guifg=Blue
                   links to DraculaSubtle
        Last set from ~/.dotfiles/vim/_plugins/dracula/colors/dracula.vim
```

So yeah, it looks like `NonText` is either getting unset or the [pre-check condition](https://github.com/dracula/vim/blob/master/colors/dracula.vim#L28-L30) in the theme is being satisfied, so nothing gets set.

I snagged that little escape hatch from the [gruvbox theme](https://github.com/morhetz/gruvbox/) because it is extremely popular and has been actively maintained for 10+ years. Figured because they've been around so long, there was a reason why that check was there so I opted to keep it.

To confirm my suspicion, can you just enter `:hi` and scroll the list to see if you can locate one or more groups with the name `Dracula*` in existence. If so, it's not the escape hatch. If so, it is.

Thanks for your patience :pray:

**Comment by hacker1db** [comment]
![screen shot 2018-04-09 at 9 25 46 am](https://user-images.githubusercontent.com/6212228/38509871-06a161d0-3bd8-11e8-85b4-7229775b35bc.png) Sadly, I think its something else.. Thanks for all the help.

**Comment by benknoble** [comment]
What about these lines ? (I think it's [73-ish][hacker1db])

```
" make the highlighting of tabs and other non-text less annoying
highlight SpecialKey ctermbg=none ctermfg=8
highlight NonText ctermbg=none ctermfg=8
" make comments and HTML attributes italic
highlight Comment cterm=italic
highlight htmlArg cterm=italic
```

[hacker1db]: https://github.com/hacker1db/Dotfiles/blob/dev_branch/config/nvim/init.vim#L73

**Comment by dsifford** [comment]
@benknoble found it.. That's the issue for sure. Remove all those 👍 👏

**Comment by dsifford** [comment]
@nineKnight Here's a reminder if you forgot how git works:

```sh
git checkout 3eaed4ff9fe8441bee02a00fbe80ac12cffd9b1c
```

Feel free to stay detached in the past if the updates don't suit your needs. We're using "garbage" patches because the terminal combination of every conceivable user is not something we can control. So the argument doesn't really apply here, but thanks.

**Comment by nineKnight** [comment]
@dsifford I don't even know the feature exists before I find this issue and your commit 52d955a. You maintainers never change vim bg these years, but these days. Do you know how we feel when our vim suddenly changed by unknown who, what or where? Do you want everyone who suffers this feature get into the source code, even its history? Come on, ridiculous.

Agressive changes should be tested completely.
New features should be declared.
New unnecessary features should be disabled by default.

I'm sorry if words hurt you. Except this one, I appreciate your works.

**Comment by hacker1db** [comment]
@dsifford  and @benknoble  sorry for delayed response, that commented out those lines and it fixed the issue! 👍 
Thank you for all of your help!

**Comment by dmitriyK1** [comment]
What helped me is setting these variables in .vimrc:

```viml
let &t_ZH="\e[3m"
let &t_ZR="\e[23m"
```

**Comment by bangarangler** [comment]
larshisken answer totally fixed my vim in tmux/iterm 2.  I had almost given up and was just getting used to the gray.  This gave me that beautiful black background in vim instantly!  doubt anyone is still checking this old thread but fuck yes!  much appreciated ; ))

**Comment by zkity** [comment]
> Setting `let g:dracula_colorterm=0` before calling `colorscheme dracula` solves the issue for me.
> 
> "And I really don't want to see these garbage patches just mindlessly sent around." ---- Linus Torvalds
> 
> I think this feature is one of them.

You really should to try this, it's nice for me!

**Comment by mr-pmillz** [comment]
Yep! This was driving me bonkers for a minute!

After much time wasted going down too many rabbit holes trying to get the dracula theme to work as intended, (Amazing theme by the way, Much thanks and respect!) What also fixed it for me was declaring:
`let dracula_colorterm = 0` 
and 
`let dracula_italic = 0`

**BEFORE** calling `colorscheme dracula` 

Here's my working config snippets in relevant order:
This is in `~/.vim_runtime/my_configs.vim` using the https://github.com/amix/vimrc project for awesome customized vim.

```c
try
    " vim-devicons set encoding utf-8
    set encoding=UTF-8
    syntax enable
    filetype plugin indent on

    " Set Dracula to the default Theme
    let g:dracula_colorterm = 0
    let g:dracula_italic = 0
    highlight Normal ctermbg=None
    colorscheme dracula
    " Show Line Numbers
    set nu
    " Set Background to dark mode
    " For Light mode, switch color to peaksea and set background=light
    " set background=light
    set background=dark
    " <Snippet>

    " vim-airline settings
    " DO NOT USE badwolf airline_theme!!! This seemed to break stuff
    " let g:airline_theme = 'badwolf'
    let g:airline_theme = 'dracula'
    " ....more airline customizations but not relevant to this post... </Snippet>
    
catch
endtry
```
